WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift

```diff
@@ -90,3 +107,78 @@ enum CardPresentPaymentAdaptedCollectOrderPaymentResult {
     case success
     case cancellation
 }
+
+private extension CardPresentPaymentCollectOrderPaymentUseCaseAdaptor {
+    func onCancel(paymentEventDetails: CardPresentPaymentEventDetails) {
+        switch paymentEventDetails {
+            case .scanningForReaders(let endSearch):
+                endSearch()
+            case .scanningFailed(_, let endSearch):
+                endSearch()
+            case .bluetoothRequired(_, let endSearch):
+                endSearch()
+            case .connectingToReader:
+                // TODO: cancel connection if possible?
+                return
+            case .connectingFailed(_, _, let endSearch):
+                endSearch()
+            case .connectingFailedNonRetryable(_, let endSearch):
+                endSearch()
+            case .connectingFailedUpdatePostalCode(_, let endSearch):
+                endSearch()
+            case .connectingFailedChargeReader(_, let endSearch):
+                endSearch()
+            case .connectingFailedUpdateAddress(_, _, let endSearch):
+                endSearch()
+            case .foundReader(_, _, _, let endSearch):
+                endSearch()
+            case .updateProgress(_, _, let cancelUpdate):
+                cancelUpdate?()
+            case .updateFailed(_, let cancelUpdate):
+                cancelUpdate()
+            case .updateFailedNonRetryable(let cancelUpdate):
+                cancelUpdate()
+            case .updateFailedLowBattery(_, let cancelUpdate):
+                cancelUpdate()
+            case .preparingForPayment(cancelPayment: let cancelPayment):
+                cancelPayment()
+            case .tapSwipeOrInsertCard(_, cancelPayment: let cancelPayment):
+                cancelPayment()
+            case .paymentSuccess(done: let done):
+                done()
+            case .paymentError(_, _, cancelPayment: let cancelPayment):
+                cancelPayment()
+            case .paymentErrorNonRetryable(_, cancelPayment: let cancelPayment):
+                cancelPayment()
+            case .processing:
+                cancelPayment()
+            case .displayReaderMessage:
+                cancelPayment()
+            /// An alert to notify the merchant that the transaction was cancelled using a button on the reader
+            case .cancelledOnReader:
+                cancelPayment()
+            /// Before reader connection
+            case .selectSearchType:
+                cancelReaderSearch()
+            /// Connection already completed, before attempting payment
+            case .validatingOrder:
+                cancelPayment()
```

Comment:
From testing, at the time of `cancelPayment` (canceling Payment Intent (PI)) at this state, the PI could already be created and thus an error was thrown from canceling the PI. This also results in the payment state still advancing to the `tapSwipeOrInsertCard` state. We need to make sure the PI is recreated (since updating PI isn't an option pdfdoF-54W-p2) when the cart is updated, after the order sync integration is merged.

Reply:
We could potentially move the `invalidated` check to right before `attemptPayment` is called in the use case, which might improve this...

I suspect it's that the payment has not actually begun when we call `cancelPayment`, so it does nothing, but the order check succeeds and _then_ the payment starts.